WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb

```diff
@@ -0,0 +1,111 @@
+        RV::ValidationWarning.new(:events, competition2.id,
+                                  ERV::CHOOSE_MAIN_EVENT_WARNING),
+        RV::ValidationWarning.new(:events, competition2.id,
+                                  ERV::MISSING_RESULTS_WARNING,
+                                  event_id: "555"),
+      ]
+
+      expected_errors = [
+        RV::ValidationError.new(:events, competition2.id,
+                                ERV::UNEXPECTED_RESULTS_ERROR,
+                                event_id: "444"),
+      ]
+
+      validator_args.each do |arg|
+        erv = ERV.new.validate(arg)
+        expect(erv.warnings).to match_array(expected_warnings)
+        expect(erv.errors).to match_array(expected_errors)
+      end
+    end
+
+    it "triggers rounds-related errors and warnings" do
+      # Triggers:
+      # CHOOSE_MAIN_EVENT_WARNING
+      # UNEXPECTED_ROUND_RESULTS_ERROR
+      # MISSING_ROUND_RESULTS_ERROR
+      cutoff = Cutoff.new(number_of_attempts: 2, attempt_result: 50*100)
+      # Add some rounds to trigger the rounds validation
+      FactoryBot.create(:round,
+                        competition: competition1, event_id: "333oh",
+                        total_number_of_rounds: 2)
+      # This round is added to trigger the missing round error
+      FactoryBot.create(:round,
+                        competition: competition1, event_id: "333oh",
+                        number: 2, total_number_of_rounds: 2)
+      FactoryBot.create(:round, competition: competition2, event_id: "555")
+
+      [Result, InboxResult].each do |model|
+        result_kind = model.model_name.singular.to_sym
+        # Create a result over a cutoff which does not exist in rounds data
+        FactoryBot.create(result_kind, :over_cutoff,
+                          competition: competition1, eventId: "333oh",
+                          cutoff: cutoff)
+        FactoryBot.create(result_kind, competition: competition2, eventId: "555")
+        FactoryBot.create(result_kind,
+                          competition: competition2, eventId: "222", roundTypeId: "c")
+      end
+      expected_errors = [
+        RV::ValidationError.new(:rounds, competition1.id,
+                                ERV::UNEXPECTED_COMBINED_ROUND_ERROR,
+                                round_name: "3x3x3 One-Handed Final"),
+        RV::ValidationError.new(:rounds, competition1.id,
+                                ERV::MISSING_ROUND_RESULTS_ERROR,
+                                round_id: "333oh-1"),
+        RV::ValidationError.new(:rounds, competition2.id,
+                                ERV::UNEXPECTED_ROUND_RESULTS_ERROR,
+                                round_id: "222-c"),
+      ]
+
+      expected_warnings = [
+        RV::ValidationWarning.new(:events, competition1.id,
+                                  ERV::CHOOSE_MAIN_EVENT_WARNING),
+        RV::ValidationWarning.new(:events, competition2.id,
+                                  ERV::CHOOSE_MAIN_EVENT_WARNING),
```

Comment:
Both test examples trigger CHOOSE_MAIN_EVENT_WARNING -- should there be an example that doesn't, to make sure it is not triggered when it shouldn't be?

Reply:
I've added that!